ktor-client/ktor-client-core/common/src/io/ktor/client/features/DefaultRequest.kt

```diff
@@ -37,3 +37,49 @@ public fun HttpClientConfig<*>.defaultRequest(block: HttpRequestBuilder.() -> Un
         block()
     }
 }
+
+/**
+ * Set the [baseUrl] to use it in sub-requests.
+ * To override the existing [baseURL] in a sub-request, the full [URL] is needed in the [HttpRequestBuilder].
+ *
+ * The given [baseUrl] cannot have a query or a fragment part.
+ * The given [baseUrl] cannot be the default [URLBuilder].
+ */
+public fun HttpRequestBuilder.baseURL(baseUrl: String) {
+    baseURL(URLBuilder().takeFrom(baseUrl))
+}
+
+/**
+ * Set the [baseUrlBuilder] to use it in sub-requests.
+ * To override the existing [baseURL] in a sub-request, the full [URL] is needed in the [HttpRequestBuilder].
+ *
+ * The given [baseUrlBuilder] cannot have a query or a fragment part.
+ * The given [baseUrlBuilder] cannot be the default [URLBuilder].
+ */
+public fun HttpRequestBuilder.baseURL(baseUrlBuilder: URLBuilder) {
+    require(baseUrlBuilder.parameters.build() == Parameters.Empty && baseUrlBuilder.fragment.isEmpty()) {
+        "The baseURL cannot have a query or a fragment"
+    }
+
+    url {
+        val defaultURLBuilder = URLBuilder()
+        val isDefault = this == defaultURLBuilder
+        val onlyPathSet = startsWith(defaultURLBuilder)
+        if(!isDefault && onlyPathSet) {
+            val requestedPath = encodedPath.removePrefix("/")
+            takeFrom(baseUrlBuilder)
+            encodedPath += if (encodedPath.endsWith("/")) {
+                requestedPath
+            } else {
+                "/$requestedPath"
+            }
+        }
+    }
+}
+
+internal fun URLBuilder.startsWith(other: URLBuilder): Boolean =
+    protocol == other.protocol
+        && host == other.host
+        && port == other.port
```

Comment:
Note that port number could be "default" or specified explicitly so it may not match

Reply:
Switched to authority, which handles this special case